src/mlpack/methods/ann/layer/multihead_attention_impl.hpp

```diff
@@ -0,0 +1,231 @@
+  outBias = arma::zeros(embedDim, 1);
+  XavierInitialization().Initialize(queryWt, embedDim, embedDim);
+  XavierInitialization().Initialize(keyWt, embedDim, embedDim);
+  XavierInitialization().Initialize(valueWt, embedDim, embedDim);
+  XavierInitialization().Initialize(outWt, embedDim, embedDim);
+}
+
+template <class AttentionPolicy, typename InputDataType,
+          typename OutputDataType>
+template <typename eT>
+void MultiheadAttention<AttentionPolicy, InputDataType, OutputDataType>::
+Forward(const arma::Mat<eT>& input, arma::Mat<eT>& output)
+{
+  const size_t batchSize = input.n_cols;
+  arma::Cube<eT> query, key, value;
+  AttentionPolicy::splitInput(input, query, key, value, tgtSeqLen, srcSeqLen,
+      embedDim);
+
+  for (size_t i = 0; i < input.n_cols; ++i)
+  {
+    query.slice(i) = queryWt * query.slice(i)
+        + arma::repmat(queryBias, 1, tgtSeqLen);
+    key.slice(i) = keyWt * key.slice(i) + arma::repmat(keyBias, 1, srcSeqLen);
+    value.slice(i) = valueWt * key.slice(i)
+        + arma::repmat(valueBias, 1, srcSeqLen);
+  }
+  query /= std::sqrt(headDim);
+
+  if (!attnMask2d.is_empty())
+  {
+    if (attnMask2d.n_rows != tgtSeqLen || attnMask2d.n_cols != srcSeqLen)
+      Log::Fatal << "The size of the 2D `attn_mask` is not correct."
+                 << std::endl;
+  }
+  else if (!attnMask3d.is_empty())
+  {
+    if (attnMask3d.n_rows != tgtSeqLen || attnMask3d.n_cols != srcSeqLen ||
+        attnMask3d.n_slices != batchSize * numHeads)
+      Log::Fatal << "The size of the 3D `attn_mask` is not correct."
+                 << std::endl;
+  }
+
+  if (!keyPaddingMask.is_empty())
+  {
+    if (keyPaddingMask.n_rows != srcSeqLen ||
+        keyPaddingMask.n_cols != batchSize)
+        Log::Fatal << "The size of the `keyPaddingMask` is not correct."
+                   << std::endl;
+  }
+
+  query = arma::Cube<eT>(query.memptr(), headDim, tgtSeqLen,
+      numHeads * batchSize, 0);
+  key = arma::Cube<eT>(key.memptr(), headDim, srcSeqLen,
+      numHeads * batchSize, 0);
+  value = arma::Cube<eT>(value.memptr(), headDim, srcSeqLen,
+      numHeads * batchSize, 0);
+  arma::Cube<eT> Wt3d(srcSeqLen, tgtSeqLen, numHeads * batchSize);
+
+  for (size_t i = 0; i < numHeads; ++i)
+  {
+    Wt3d.slice(i) = key.slice(i).t() * query.slice(i);
+  }
+
+  if (!attnMask2d.is_empty())
+    Wt3d.each_slice() += attnMask2d;
+  else if (!attnMask3d.is_empty())
+    Wt3d += attnMask3d;
+
+  if (!keyPaddingMask.is_empty())
+    for (size_t i = 0; i < Wt3d.n_slices; ++i)
+    {
+      Wt3d.slice(i) += arma::repmat(keyPaddingMask, 1, tgtSeqLen);
+    }
+
+  arma::Mat<eT> Wt2d = arma::Mat<eT>(Wt3d.memptr(), srcSeqLen * tgtSeqLen,
+      numHeads * batchSize, false, false);
+
+  boost::apply_visitor(ForwardVisitor(Wt2d, boost::apply_visitor(
+      outputParameterVisitor, softmaxModule)), softmaxModule);
+  boost::apply_visitor(ForwardVisitor(boost::apply_visitor(
+      outputParameterVisitor, softmaxModule), Wt2d), dropoutModule);
```

Comment:
Do you need a visitor here? Why don't you call `softmaxModule->Forward()` directly?

Reply:
I'm clear about that but I think we will need the output of the softmax module for backward propagation. I think the visitor does that.